### PR TITLE
Hide group header and header separator if no tasks are left after fitering

### PR DESCRIPTION
**Pull request by cslux**
This fixes my reported issue #56. #61 is very likely to be solved with this as well.

### CODE REVIEW

**Comment by alphapapa** [comment]
Hi Christian,

This looks very interesting.  Thanks for your work on it.  I hardly use agenda filters myself, so I'm not very familiar with them.  A few thoughts:

1.  Would you mind writing a few sentences summarizing the changes in the code?  i.e. explaining why and how each function was changed, at a high level.
2.  I think there might be a simpler way to handle hiding and showing text when filters are modified.  For example, something like:
    1.  Using `before` function advice, before Org Agenda filters are applied, remove all `invisible` text-properties from the buffer.
    2.  Using `after` function advice, after the Org Agenda filters are applied, apply `invisible` text-property to any headers which are followed by another header or end-of-buffer (which would indicate that they had no items).
  
    I think code that worked roughly that way would be easier to follow and less likely to have bugs, and I think it would obviate the `org-super-agenda-show-line` function, making the whole patch simpler.
3.  Nitpick: There seems to be an unintended indentation change in the minor mode docstring.
4.  Finally, since this involves advising more functions, I think it should be tested very carefully before releasing, ideally by multiple users.  At this stage of `org-super-agenda`'s lifetime, my top priority is to not break anything or add new bugs, as one of the things I value most in Org is stability, and I think other users do as well.

Thanks.

**Comment by cslux** [comment]
Hey Alphapapa!

> 1 Would you mind writing a few sentences summarizing the changes in the code? i.e. explaining why and
>  how each function was changed, at a high level.

Sure.

> 2 I think there might be a simpler way to handle hiding and showing text when filters are modified. For
>  example, something like:
>
>  1 Using before function advice, before Org Agenda filters are applied, remove all invisible text-properties
>  from the buffer.
>  2 Using after function advice, after the Org Agenda filters are applied, apply invisible text-property to any
>  headers which are followed by another header or end-of-buffer (which would indicate that they had no
>  items).
>
>  I think code that worked roughly that way would be easier to follow and less likely to have bugs, and I
>  think it would obviate the org-super-agenda-show-line function, making the whole patch simpler.

Well, I first tried to implement something going top-down, but that was actually
more complicated than bottom-up, as in this case, due to
`org-super-agenda-header-separator`. `org-super-agenda-header-separator` could
be anything. With the bottom-up method, we can say, If `org-super-agenda-header`
will be invisible, any followed text with the property
`org-super-agenda-header-separator` needs to be invisible as well.

It very likely could have been the case that my brain played nifty games with
me.

I will revisit the algo, and might remove any filter we have applied before (Or
do you mean any, any, i.e also those set by org itself).

> 3 Nitpick: There seems to be an unintended indentation change in the minor mode docstring.

Ah, sorry! I use aggressive-indent-mode. Will fix that.

> 4 Finally, since this involves advising more functions, I think it should be tested very carefully before
>  releasing, ideally by multiple users. At this stage of org-super-agenda's lifetime, my top priority is to not
>  break anything or add new bugs, as one of the things I value most in Org is stability, and I think other
>  users do as well.

No problem! In the meantime I will load `org-super-agenda` from source.

Christian

**Comment by cslux** [comment]
When the group header name is over multiple lines the current implementation produces an incorrect result.

**Comment by cslux** [comment]
@alphapapa I improved it a bit. It works now with multiline group header/separator as well.

**Comment by cslux** [comment]
It got actually more code, but  it is more readable IMO.

**Comment by cslux** [comment]
Hmmm, somehow the header separator "\n" makes trouble.

**Comment by cslux** [comment]
@alphapapa This version is much better :-)  3d9e263

**Comment by alphapapa** [comment]
That does look better, thanks.  A couple of thoughts:

1.  I encourage you to use `cl-loop` for the loop, as it can result in clearer and more concise code (IMO, anyway; some people hate it).  For example, it has clauses like `while` and `until`, and you can use `return` to break the loop.  It's documented in the `cl` info manual in Emacs, and there are lots of examples in this package.
2.  In the first part of the patch, `org-super-agenda--make-agenda-header` is modified.  Would you please try to reduce the changes in that section to a minimum, or explain why the changes were made?  e.g. I don't understand why the order of the strings that are concatenated is changed.
3.  For the additional text property, please use `org-super-agenda-header` rather than `org-super-agenda`.

Thanks for your work on this.

**Comment by cslux** [comment]
>     1. I encourage you to use `cl-loop` for the loop, as it can result in clearer and more concise code (IMO, anyway; some people hate it).  For example, it has clauses like `while` and `until`, and you can use `return` to break the loop.  It's documented in the `cl` info manual in Emacs, and there are lots of examples in this package.

Ok, done!

>     2\. In the first part of the patch, `org-super-agenda--make-agenda-header` is modified.  Would you please try to reduce the changes in that section to a minimum, or explain why the changes were made?  e.g. I don't understand why the order of the strings that are concatenated is changed.

I changed it back, and added a `org-add-props` before the final `concat`.

>     3\. For the additional text property, please use `org-super-agenda-header` rather than `org-super-agenda`.

Done!

Thanks for the guidance, I learnt quite a bit as a result of your review.

**Comment by cslux** [comment]
Just discovered that the "Timed Items" header group is a special case and should never be hidden.

**Comment by alphapapa** [comment]
Hi,

Sorry for neglecting this PR.  Feel free to ping me in case I forget about it.

> Just discovered that the "Timed Items" header group is a special case and should never be hidden.

1.  Why is that?
2.  Assuming that's correct, is the change in 9c0d58a the right thing to do, i.e. is it right to not check `type`?  Should we check both?

Thanks for your work on this.

**Comment by cslux** [comment]
Hey @alphapapa!

No problem!

1. It can be toggled with `org-agenda-log-mode`.
2. With 'type the "Timed items" header gets hidden but not the time **grid**.

But, I don't feel good about it,  since it behaves different than the rest. I will lock over it again, and hide the "Log items" and "Timed items" group as well, when there is no task visible.

**Comment by cslux** [comment]
@alphapapa 

I reverted the last commit, and improved it. Any group gets now hidden when there is no visible task.

**Comment by cslux** [comment]
@alphapapa Ping! :-)

**Comment by alphapapa** [comment]
Sorry for neglecting this.  I'll try to look at it again soon.

**Comment by cslux** [comment]
@alphapapa Please review and merge it. I have squashed alls commits to a single commit and has no conflicts with master.

**Comment by alphapapa** [comment]
@cslux Looking at it again, I don't know about this.  I don't like adding all that function advice, even though it may be necessary to implement this feature.  I'm not sure this feature is important enough to justify doing that.  

Regarding the code itself, the function is not formatted idiomatically, and the flow of the code is hard to follow because of the nested loops and the way variables are used.  There are also some spelling mistakes in the comments (`flyspell-prog-mode` may be helpful).  So I don't want to merge it in its current state.  

If you really want this to be merged, I would want the function to be easier to follow, e.g. maybe you could use `cl-flet` or `cl-labels` to write some well-named and well-defined functions within the function.  For example, in this code, I used `cl-macrolet` to make the code easier to follow than if the bodies of the macrolets had all been inlined: https://github.com/alphapapa/magit-todos/blob/master/magit-todos.el#L537  

However, I won't promise to merge it in advance, so if you're not interested in working on it further, we can leave it here, and you can keep using it in your own fork, and maybe one of us can work on it in the future.

Thanks.

**Comment by cslux** [comment]
I'll keep a fork than. Feel free to close it!

**Comment by cslux** [comment]
@alphapapa I pushed something (c4a62df) which uses `cl-flet`. Please have look and tell me what you think. Also, I could remove one advice by advicing `org-agenda-finalize` instead of `org-agenda-redo`.

Regarding your preferred formatting, please tell me what settings do you use.

**Comment by alphapapa** [comment]
Hi again,

Thanks for your patience.  I looked at the current code, and that does look much easier to understand!  I'll submit a few small suggestions as review comments.

I want to be honest: at this time, I'm still not promising to merge it, because I'd like to think about it and test it some more.  But assuming that there are no performance regressions due to the extra advices and such, I think this is looking reasonable.  :)

Thanks for your work on this.

File: org-super-agenda.el [hunk 1]
```diff
@@ -272,12 +272,15 @@ face `org-super-agenda-header' appended, and the text properties
...
-         (org-add-props s nil
+         (org-add-props s nil 'face 'org-super-agenda-header
            'keymap org-super-agenda-header-map
            ;; NOTE: According to the manual, only `keymap' should be necessary, but in my
            ;; testing, it only takes effect in Agenda buffers when `local-map' is set, so
            ;; we'll use both.
-           'local-map org-super-agenda-header-map)
+           'local-map org-super-agenda-header-map
+           'org-super-agenda-header t)
+         (org-add-props org-super-agenda-header-separator nil
```
Comment (alphapapa) [inline org-super-agenda.el]: It's probably not quite correct to add this property to this special variable every time this function is called.  I think the property should be applied to `separator` instead.  However, see next comment...

File: org-super-agenda.el [hunk 2]
```diff
@@ -272,12 +272,15 @@ face `org-super-agenda-header' appended, and the text properties
...
+           'org-super-agenda-header t)
          ;; Don't apply faces and properties to the separator part of the string.
          (concat separator s)))))
```
Comment (alphapapa) [inline org-super-agenda.el]: Maybe, instead of using `org-add-props` on `separator` and `s`, we should simply use `propertize` on this call to `concat`.  Or would the keymaps be a problem?  What do you think?

File: org-super-agenda.el [hunk 3]
```diff
@@ -302,15 +305,27 @@ marker."
...
+                                 (lambda (from fn)
+                                   ;; Disable mode
+                                   (advice-remove from fn)))))
+    (funcall advice-function-filter-return #'org-agenda-finalize-entries
              #'org-super-agenda--filter-finalize-entries)
+    (funcall advice-function-after #'org-agenda-filter-apply
+             #'org-super-agenda--hide-or-show-groups)
+    (funcall advice-function-after #'org-agenda-finalize
+             #'org-super-agenda--hide-or-show-groups)
+
```
Comment (alphapapa) [inline org-super-agenda.el]: I would prefer to reduce the number of advices to the minimum, if possible.  I haven't studied the code in question, so it may not be possible to reduce this further.  What do you think?

File: org-super-agenda.el [hunk 4]
```diff
@@ -1099,6 +1114,53 @@ actually the ORDER for the groups."
        (-remove #'s-blank-str? it)
        (s-join "\n" it)))
 
+;;;; Hide/Show filter
+(defun org-super-agenda--hide-or-show-groups (&rest _)
+  "Hide/Show any empty/non-empty groups after `org-agenda-finalize',
+`org-agenda-filter-apply'  was called."
+  (cl-labels ((header-p () (org-get-at-bol 'org-super-agenda-header))
+              (grid-p () (not (or  (org-get-at-bol 'org-agenda-structural-header)
```
Comment (alphapapa) [inline org-super-agenda.el]: Nitpick: extra space after `or`.  :)

Comment (alphapapa) [inline org-super-agenda.el]: It might be better to get the list of text properties at that position and compare it against a list of properties, rather than calling `org-get-at-bol` 4 times.  I'm not sure which would be faster.  Would you be willing to benchmark that?

File: org-super-agenda.el [hunk 5]
```diff
@@ -1099,6 +1114,53 @@ actually the ORDER for the groups."
...
+                                   (org-get-at-bol 'org-agenda-date-header)
+                                   (org-get-at-bol 'type)
+                                   (org-get-at-bol 'org-super-agenda-header))))
+              (group-item-visible-p () (and (org-get-at-bol 'type) (not (org-get-at-bol 'invisible))))
+              (next-header () (let (header nohide grid-end)
+                                (while (not (or (bobp) header))
+                                  (cond
```
Comment (alphapapa) [inline org-super-agenda.el]: Nitpick: It's a matter of style, but for consistency with the rest of the code, please put the first cond on the same line as `cond`.

File: org-super-agenda.el [hunk 6]
```diff
@@ -1099,6 +1114,53 @@ actually the ORDER for the groups."
...
+                                    ((header-p)
+                                     (setq header (list (1- (or (previous-single-property-change (point-at-eol) 'org-super-agenda-header) (1+ (point-min))))
```
Comment (alphapapa) [inline org-super-agenda.el]: That line is really long.  Could the second part of the `or` go on another line?  :)

File: org-super-agenda.el [hunk 7]
```diff
@@ -1099,6 +1114,53 @@ actually the ORDER for the groups."
...
+                                                        (or grid-end (point-at-eol))
+                                                        nohide)))
+                                    ((group-item-visible-p)
+                                     (setq nohide t))
```
Comment (alphapapa) [inline org-super-agenda.el]: Another nitpick: would it make the code clearer if it were `hide-p` rather than `nohide`?  Of course that would invert its meaning, but it might be worth it.  What do you think?

File: org-super-agenda.el [hunk 8]
```diff
@@ -1099,6 +1114,53 @@ actually the ORDER for the groups."
...
+                                    ((and (grid-p) (not grid-end))
+                                     (setq grid-end (point-at-eol))))
+                                  (beginning-of-line 0))
+                                header))
+              (hide-or-show-header (header)
+                (when header
+                  (cl-loop
+                     with start = (nth 0 header)
```
Comment (alphapapa) [inline org-super-agenda.el]: These three variables can be destructured with `cl-loop`, like:
```el
with (start end nohide) = header
```

File: org-super-agenda.el [hunk 9]
```diff
@@ -1099,6 +1114,53 @@ actually the ORDER for the groups."
...
+                     with end = (nth 1 header)
+                     with nohide = (nth 2 header)
+                     with props = `(invisible org-filtered org-filter-type org-super-agenda-header)
+                     initially do (goto-char end)
+                     while (and start (> (point) start))
+                     do
```
Comment (alphapapa) [inline org-super-agenda.el]: Nitpick: Please put the `when` on the same line as the `do`.

File: org-super-agenda.el (hunk 7 above)
Comment (cslux) [reply org-super-agenda.el]: Man, I thought I'm fussy :-D. If I show that my working colleagues they hopefully change there mind about me.

File: org-super-agenda.el (hunk 4 above)
Comment (cslux) [reply org-super-agenda.el]: Whoops!

File: org-super-agenda.el (hunk 3 above)
Comment (cslux) [reply org-super-agenda.el]: As far as I know, it's not possible to further reduce it. But I could/will ask that question on the org-devel mailing list.

File: org-super-agenda.el (hunk 2 above)
Comment (cslux) [reply org-super-agenda.el]: I can still call any org-agenda-mode-map bindings on the seperator.

File: org-super-agenda.el (hunk 4 above)
Comment (cslux) [reply org-super-agenda.el]: Hmm, not sure if this is the correct way to benchmark it!?

``` emacs-lisp
(benchmark 10000000 (not (let ((props (text-properties-at (point-at-bol))))
                           (or (member 'org-agenda-structural-header props)
                               (member 'org-agenda-date-header props)
                               (member 'type props)
                               (member 'org-super-agenda-header props)))))
"Elapsed time: 0.070366s"

"Elapsed time: 0.069554s"

"Elapsed time: 0.088570s"

"Elapsed time: 0.063170s"


(benchmark 10000000 (not (or (org-get-at-bol 'org-agenda-structural-header)
                             (org-get-at-bol 'org-agenda-date-header)
                             (org-get-at-bol 'type)
                             (org-get-at-bol 'org-super-agenda-header))))
"Elapsed time: 0.064548s"

"Elapsed time: 0.056398s"

"Elapsed time: 0.052549s"

"Elapsed time: 0.050331s"
```

Thanks!

**Comment by distefam** [comment]
FWIW, I'm particularly interested in this change as I commonly group my `TODO`s by project (parent) and filter by tag. This tends to leave a large number of project headings with no tasks.

File: org-super-agenda.el (hunk 4 above)
Comment (alphapapa) [reply org-super-agenda.el]: Not a large difference, but `org-get-at-bol` is faster.  Good to know.  Thanks.

Comment (alphapapa) [reply org-super-agenda.el]: Actually, I just remembered that your benchmark doesn't use `lexical-binding`, which may make a difference.  Please try with the `bench-multi-lexical` macro from https://github.com/alphapapa/emacs-package-dev-handbook.

File: org-super-agenda.el (hunk 7 above)
Comment (alphapapa) [reply org-super-agenda.el]: A wise man once said, "If you take care of the little things, the big things will take care of themselves."  ;)

**Comment by alphapapa** [comment]
@distefam Understood.  :)  If you have time to test his PR "in anger" for a little while, your feedback would be appreciated.  Since this isn't a feature I use much (or at all), it's hard for me to anticipate any issues with its implementation.

File: org-super-agenda.el (hunk 4 above)
Comment (cslux) [reply org-super-agenda.el]:
``` emacs-lisp
(bench-multi-lexical :times 500000
  :ensure-equal t
  :forms (("test-properties-at" (not (let ((props (text-properties-at (point-at-bol))))
                                       (or (member 'org-agenda-structural-header props)
                                           (member 'org-agenda-date-header props)
                                           (member 'type props)
                                           (member 'org-super-agenda-header props)))))
          ("org-get-at-bol" (not (or (org-get-at-bol 'org-agenda-structural-header)
                                     (org-get-at-bol 'org-agenda-date-header)
                                     (org-get-at-bol 'type)
                                     (org-get-at-bol 'org-super-agenda-header))))))
```
Result:

 Form | x faster than next | Total runtime |  # of GCs | Total GC runtime 
 ----- | ---------------------- | ----------------- | ------------ | -----------------------
 test-properties-at | 2.39 | 0.150025 | 0 | 0 
 org-get-at-bol | slowest | 0.359302 | 0 | 0 


Seems like `org-get-at-bol` is now slower. Hmmm, interesting :-

Comment (cslux) [reply org-super-agenda.el]: Ok, it uses `text-properties-at` now.

File: org-super-agenda.el (hunk 3 above)
Comment (cslux) [reply org-super-agenda.el]: I didn't get any response!

**Comment by cslux** [comment]
@alphapapa I was able to reduce it to just one advice. Unfortunately, this one is required unless org-mode gets a little modified.

**Comment by koleiskiy** [comment]
@cslux Thank you for your work! This PR adds a feature I wish `org-super-agenda` had because I use `org-agenda-filter` a lot.

Sadly, I found a bug which keeps empty timelines hidden after `org-agenda-remove-filter`. 

Reproducible steps are:
1. Call this function: 

```
(defun org-super-agenda ()
  (interactive)
  (org-super-agenda-mode +1)
  (let ((org-agenda-start-with-log-mode t)
        (org-agenda-span 1)
        (org-super-agenda-groups '((:time-grid t))))
    (org-agenda-list)))
```
2. Call `org-agenda-filter-by-regexp`.
3. Enter a regular expression which filters out all items.
4. Call `org-agenda-remove-filter`.

And again, thank you for the work :)

**Comment by cslux** [comment]
@koleiskiy Indeed, thanks! b9d5b3e should have fixed it.

**Comment by koleiskiy** [comment]
@cslux 
Thank you! it fixes the issue :)

@alphapapa 
I've been using @cslux's fork for a week and only encountered the mentioned bug, and not it was fixed. It would be nice if this PR could be merged. 

But before that, this feature should have a custom variable something like `org-super-agenda-hide-empty-headers-when-filtered` because I think there are some users who like the current behavior which shows empty headers when filtered. What do you think?

**Comment by alphapapa** [comment]
Hi,

Apologies for neglecting this for a while.  A few things:

1.  Looking at the diff, it appears that it needs to be rebased on `master`.  @cslux Would you be willing to do that?
2.  @cslux @koleiskiy @distefam After it's rebased, would you be willing to give the branch a quick test?  If so, please be sure to specify what versions of Emacs and Org that you're using.

If those items are done and it tests cleanly, then I have one more request before merging this: I'd like it to be gated on an additional option.  My concern here is mainly about performance in large results buffers, e.g. ones with thousands of lines (which can happen when I search all of my Org files for a common word like "Emacs").  For users who don't use this feature, it would be good to avoid scanning the buffer for things to hide/show every time.

Thanks.

**Comment by cslux** [comment]
Hello :-),

didn't expect to here again from you!

I can do that, so no merging, just a clean linear history, no problem.

Regarding the performance concern, my org-agenda files contain together
~ 35000 lines, but I have no idea how to make a performance test. Help needed!

- Chris

**Comment by cslux** [comment]
Rebase done! Also fixed a linting error in the doc-string.

**Comment by alphapapa** [comment]
Hi,

Thanks for rebasing.

I don't expect you to write a performance test.  I'd just like this feature to be gated behind an option so it can be completely disabled in case some users have performance issues.  :)

**Comment by cslux** [comment]
@alphapapa I added a customizable option to enable/disable hiding of empty group headers. Disabled by default. b89f835

**Comment by alphapapa** [comment]
Hi Christian,

Thanks for your patience on this.  I didn't intend to leave it for so long, but I only work on some packages from time to time.

I made a few small changes and merged it to master.  Thanks for putting this together.

BTW, if you submit any more PRs in the future, it's best to put them in a feature branch in your fork rather than using its master branch.

**Comment by cslux** [comment]
> BTW, if you submit any more PRs in the future, it's best to put them in a feature branch in your fork rather than using its master branch.

Meanwhile I have learned that it certainly makes sense to do so :-P
